### PR TITLE
Test document indexation

### LINKED ISSUE #8: Test that the documents were correctly indexed
Make a test to verify that the documents sent to MeiliSearch were correctly indexed

### PR DESCRIPTION
**Pull request by mdubus**
Closes #8 

## What's inside this PR ? 

- Validate function is now a module (in order to be imported in the tests). The watch mode has been updated accordingly
- Addition of jest as dependency, it's configuration file, and allowing of `.env` file
- Addition of a file `validate-options.test.js` testing the validation of the options
- Addition of a file `index-to-meilisearch.test.js` testing the logic of the plugin
- Update of the scripts inside the `package.json`
- Update of `CONTRIBUTING.md` for further explaining of the project's testing
- Update of the CI

## How to test ?

You should have a MeiliSearch and the front's graphQL server running in order to run the tests.
You can do so with the following commands: 

```bash
docker run -it --env MEILI_MASTER_KEY=masterKey -p 7700:7700 getmeili/meilisearch:v0.21.0rc6
```

Then, you can run the tests:

`yarn test`

or: 

`yarn test:watch`

## Screenshots

Capture d’écran 2021-08-02 à 10 25 28

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -38,19 +38,26 @@ yarn
...
+#### Linter
+
+```bash
+# Linter
+yarn lint
+# Linter with fixing
+yarn lint:fix
+```
+
+#### End-to-End tests
```
Comment (bidoubiwa) [inline CONTRIBUTING.md:50]: End to end tests,integrations tests and linting should not be separated into two categories. It avoid having a contributing guide that is overwhelming. 


They can be added in the same code sample: 

```bash
# Run a MeiliSearch instance
docker pull getmeili/meilisearch:latest # Fetch the latest version of MeiliSearch image from Docker Hub
docker run -p 7700:7700 getmeili/meilisearch:latest ./meilisearch --master-key=masterKey --no-analytics=true

# Tests the project
yarn test # integration tests
yarn test:e2e # end to end tests

# Tests the project in watch/open mode
yarn test:watch
yarn test:e2e:watch

# Linter
yarn lint
# Linter with fixing
yarn lint:fix
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -60,6 +67,25 @@ yarn cy:run # Run all tests
...
+
+Some environement variables are needed in order to run this tests: `MEILI_HTTP_ADDR`, `MEILI_MASTER_KEY`, and `MEILI_INDEX_NAME`. It is possible to provide them in a `.env` file inside the `tests` folder.
+
+Example of `.env`:
+
+```bash
+MEILI_HTTP_ADDR="http://localhost:7700"
+MEILI_MASTER_KEY="masterKey"
+MEILI_INDEX_NAME="myIndex"
+```
```
Comment (bidoubiwa) [inline CONTRIBUTING.md:87]: Environment variables for testing should have a default value a bit like on the other SDK's (example)[https://github.com/meilisearch/meilisearch-js/blob/main/tests/meilisearch-test-utils.ts].

You can provide the file name where the user can change these variables if he'd like to but it is not necessary per se.

File: package.json [hunk 3]
```diff
@@ -45,8 +50,11 @@
     "eslint-plugin-cypress": "^2.11.3",
     "eslint-plugin-prettier": "^3.4.0",
     "eslint-plugin-react": "^7.24.0",
+    "gatsby": "^3.10.2",
+    "jest": "^27.0.6",
     "prettier": "^2.3.2",
-    "react": "^17.0.2"
+    "react": "^17.0.2",
```
Comment (bidoubiwa) [inline package.json:56]: is `react` needed in the root dir?

**Review by bidoubiwa — CHANGES_REQUESTED** [review]
### Files should keep the same naming convention: 
ex: `indexToMeiliSearch.test.js` -> `index-to-meilisearch.test.js`
ex `article_page.spec.js` -> `article-page.spec.js`

### test redundancy

when doing 

```js
  expect(() => validate()).toThrow()
  expect(() => validate()).toThrow("...")
```

These might be redundant as they do the same thing but one time more precisely than the other.

File: CONTRIBUTING.md (hunk 2 above)
Comment (bidoubiwa) [inline CONTRIBUTING.md:87]: this did not work for me :( but see main comment of review

Comment (mdubus) [reply CONTRIBUTING.md]: Just added a fallback in the `fakeConfig` (`tests/utils.js` file) as I prefer not to push a `.env` file 🙂

File: package.json (hunk 3 above)
Comment (mdubus) [reply package.json]: Nope, you're right. Just removed it from the dependencies. 
The only inconvenience is that I now have this message on `yarn lint`:
```
Warning: React version was set to "detect" in eslint-plugin-react settings, but the "react" package is not installed. Assuming latest React version for linting.
```

Do you want me to add a `.eslintrc` into the `playground` folder with all the react plugin & configurations for eslint (and remove them from the root), or do you want to leave it like this ?

File: CONTRIBUTING.md (hunk 2 above)
Comment (mdubus) [reply CONTRIBUTING.md]: Should work now. I added the `playground:start` concurrently 😉

**Comment by mdubus** [comment]
> ### Files should keep the same naming convention:
> ex: `indexToMeiliSearch.test.js` -> `index-to-meilisearch.test.js`
> ex `article_page.spec.js` -> `article-page.spec.js`

Done

> 
> ### Test redundancy
> when doing
> 
> ```js
>   expect(() => validate()).toThrow()
>   expect(() => validate()).toThrow("...")
> ```
> 
> These might be redundant as they do the same thing but one time more precisely than the other.
> 

Done

> ### Integration tests
> `yarn test` is not possible without running the gatsby playground. This command should be runable without any further configuration. I think we should either mock the answer of graphQL or consider launching the playground concurrently when running `yarn test`.

I added the playground concurrently 👍

File: tests/index-to-meilisearch.test.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+const { fakeConfig, fakeGraphql, fakeReporter } = require('./utils')
+
+const activity = fakeReporter.activityTimer()
+
+const client = new MeiliSearch({
+  host: fakeConfig.host,
+  apiKey: fakeConfig.apiKey,
+})
+
+describe('index to MeiliSearch', () => {
```
Comment (bidoubiwa) [inline tests/index-to-meilisearch.test.js:13]:
```suggestion
describe('Index to MeiliSearch', () => {
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -38,26 +38,35 @@ yarn
 
 Each PR should pass the tests and the linter to be accepted.
 
+
```
Comment (bidoubiwa) [inline CONTRIBUTING.md:41]:
```suggestion
```

**Review by bidoubiwa — CHANGES_REQUESTED** [review]
Very small changes and we r good to go 🔥🔥

**Review by bidoubiwa — APPROVED** [review]
LGTM 🔥

**Comment by mdubus** [comment]
bors merge

**Pull request merged**